---
title: "**Measuring population differentiation with the fixation index (*F*~ST~)**"
output: html_document
---

### Understanding *F*~ST~

*F*~ST~ measures genetic differentiation between populations. It ranges from 0 (no differentiation) to 1 (complete differentiation).

Many different estimators of *F*~ST~ have been proposed. Let's first apply the simplest formulation to illustrate how it is calculated, based on some very simple data sets where we can visually assess allele frequencies. Then we will use a more complex estimator that is more commonly used for real data sets.

Let's consider three different scenarios, each with two populations being compared. In the first scenario, the two populations have identical allele compositions (AAABBB):

```{r, message=FALSE, warning=FALSE}
scenario1 <- data.frame(
  population = rep(c("Population 1", "Population 2"), each = 6),
  allele = c("A", "A", "A", "B", "B", "B", "A", "A", "A", "B", "B", "B")
)
```

In the second scenario, Population 1 has AABBBB and Population 2 has AAAABB:

```{r, message=FALSE, warning=FALSE}
scenario2 <- data.frame(
  population = rep(c("Population 1", "Population 2"), each = 6),
  allele = c("A", "A", "B", "B", "B", "B", "A", "A", "A", "A", "B", "B")
)
```

In the third scenario, the two populations have entirely different allele compositions. Population 1 has AAAAAA and Population 2 has BBBBBB:

```{r, message=FALSE, warning=FALSE}
scenario3 <- data.frame(
  population = rep(c("Population 1", "Population 2"), each = 6),
  allele = c("A", "A", "A", "A", "A", "A", "B", "B", "B", "B", "B", "B")
)
```

*F*~ST~ can be calculated as:

$$
F_{ST} = \frac{\pi_{between} - \pi_{within}}{\pi_{between}}
$$

where:  
- \( \pi_{within} \) is the average genetic diversity within populations.  
- \( \pi_{between} \) is the genetic diversity between populations.  

Define an R function to calculate *F*~ST~ using this formula and apply it to our three scenarios.

```{r, message=FALSE, warning=FALSE}
# Define the function to calculate FST
calculate_fst <- function(pop_df) {
  # Calculate allele frequencies within each population
  pop1_freq <- mean(pop_df$allele[pop_df$population == "Population 1"] == "A")
  pop2_freq <- mean(pop_df$allele[pop_df$population == "Population 2"] == "A")
  
  # Within and between population allele diversity
  pi_within <- (pop1_freq * (1 - pop1_freq) + pop2_freq * (1 - pop2_freq)) / 2
  pi_between <- (pop1_freq * (1 - pop2_freq) + pop2_freq * (1 - pop1_freq)) / 2
  
  # FST formula
  fst <- (pi_between - pi_within) / pi_between
  return(round(fst, 2))
}

# Calculate FST for each scenario
calculate_fst(scenario1)
calculate_fst(scenario2)
calculate_fst(scenario3)
```

In the first scenario, the two populations differ 0% in the genetic variation we are observing. In the second scenario, the two populations differ 20%. And in the third, they differ 100%.

In real data, we often use more complex estimators of *F*~ST~. Below we will use the 'stamppFst' function from the package 'StAMPP' to apply [Weir & Cockerham's *F*~ST~ estimator](https://doi.org/10.2307/2408641) to our real SNP data set. This estimator accounts for sample sizes, and it is one of the most widely used today.

### Load packages and data
```{r, message=FALSE, warning=FALSE, results='hide'}
# Load relevant packages
library(vcfR)
library(adegenet)
library(StAMPP)

# Set working directory
setwd("C:/Users/rafae/Downloads/aulagenpopmz/data")

# Read in SNP data
snpdata.vcf <- read.vcfR("snpdata.vcf")

# Convert vcf to genlight
snpdata.gen <- vcfR2genlight(snpdata.vcf)

# Read in sample info
sample.info <- read.table("sampleinfo.csv", sep = ",", header = TRUE)

# Reorder sample info to match order of samples in genlight
sample.info <- sample.info[match(indNames(snpdata.gen), sample.info$id), ]

# Set population membership
pop(snpdata.gen) <- sample.info[, "pop"]
```

### Calculate *F*~ST~
```{r, message=FALSE, warning=FALSE}
fst.estimates <- stamppFst(snpdata.gen, nboots = 100, percent = 95)

# See results with two decimals
round(fst.estimates$Fsts, 2)
```
