---
title: "**Identifying hybrids with triangle plots**"
output: html_document
---

### Load packages and read data
```{r, message=FALSE, warning=FALSE, results='hide'}
# Load packages
library(vcfR)
library(triangulaR)
library(ggplot2)

# Set working directory
setwd("C:/Users/rafae/Downloads/aulagenpopmz/data")

# Read in SNP data
snpdata.vcf <- read.vcfR("snpdata.vcf")

# Read in sample info
sample.info <- read.table("sampleinfo.csv", sep = ",", header = TRUE)

# Keep only relevant info in a new dataframe
popmap <- sample.info[, c("id", "pop")]

# Reorder popmap to match order of samples in vcf
popmap <- popmap[match(colnames(snpdata.vcf@gt)[-1], popmap$id), ]
```

### Prepare data and calculate indices
```{r, message=FALSE, warning=FALSE}
# Create new vcfR objects composed only of SNPs above the given allele frequency thresholds
popdiff1 <- alleleFreqDiff(vcfR = snpdata.vcf, pm = popmap, p1 = "D", p2 = "A", difference = 1)
popdiff75 <- alleleFreqDiff(vcfR = snpdata.vcf, pm = popmap, p1 = "D", p2 = "A", difference = 0.75)

# Calculate hybrid index and heterozygosity for each sample
hi.het1 <- hybridIndex(vcfR = popdiff1, pm = popmap, p1 = "D", p2 = "A")
hi.het75 <- hybridIndex(vcfR = popdiff75, pm = popmap, p1 = "D", p2 = "A")
```

### Plot results
```{r, message=FALSE, warning=FALSE}
# Plot the triangle based on fully fixed SNPs
triangle.plot(hi.het1)

# Plot the triangle based on SNPs above 0.75 fixation
triangle.plot(hi.het75)
```